### PR TITLE
BBSDEV-18209: Added support to discover VMs in a scale set

### PR DESCRIPTION
**Pull request by justinskariah**
* Added support to discover VMs in a scale set
* Added scale set to the ARM template and updated the integration test to verify scale set VMs

### CODE REVIEW

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java [hunk 1]
```diff
@@ -88,63 +81,103 @@ public void start() {
...
             LOGGER.finest("Failed to discover nodes with Azure SPI", e);
             return null;
         }
+    }
 
+    private SimpleDiscoveryNode createDiscoveryNode(int port, VirtualMachineScaleSetVM vm,
+                                                    VirtualMachineScaleSetNetworkInterface networkInterface)
+            throws UnknownHostException {
+        String privateIP = networkInterface.primaryPrivateIP();
+        if (getLocalHostAddress() != null && privateIP.equals(getLocalHostAddress())) {
```
Comment (hsch) [inline src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: Don't invoke `getLocalHostAddress` multiple times.

**Review by hsch — APPROVED** [review]

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java [hunk 2]
```diff
@@ -27,20 +27,13 @@
 import com.hazelcast.spi.partitiongroup.PartitionGroupMetaData;
 import com.microsoft.azure.CloudException;
 import com.microsoft.azure.PagedList;
-import com.microsoft.azure.management.compute.InstanceViewStatus;
-import com.microsoft.azure.management.compute.VirtualMachine;
-import com.microsoft.azure.management.compute.VirtualMachineInstanceView;
+import com.microsoft.azure.management.compute.*;
```
Comment (juan-palacios) [inline src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: It doesn't feel like wildcard imports are a thing in this code base. Maybe keep as is?

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java [hunk 3]
```diff
@@ -88,63 +81,103 @@ public void start() {
...
+
+                String primaryNetworkInterfaceId = vm.primaryNetworkInterfaceId();
+                if (primaryNetworkInterfaceId != null) {
+                    VirtualMachineScaleSetNetworkInterface networkInterface =
+                            vm.getNetworkInterface(primaryNetworkInterfaceId);
+                    nodes.add(createDiscoveryNode(port, vm, networkInterface));
+                } else {
+                    PagedList<VirtualMachineScaleSetNetworkInterface> networkInterfaces = vm.listNetworkInterfaces();
+                    if (networkInterfaces.size() > 0) {
+                        VirtualMachineScaleSetNetworkInterface networkInterface = networkInterfaces.get(0);
```
Comment (juan-palacios) [inline src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: Why the first one?

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java [hunk 4]
```diff
@@ -88,63 +81,103 @@ public void start() {
...
-            }
             LOGGER.info("Azure Discovery SPI Discovered " + nodes.size() + " nodes");
             return nodes;
         } catch (Exception e) {
             LOGGER.finest("Failed to discover nodes with Azure SPI", e);
             return null;
         }
+    }
 
+    private SimpleDiscoveryNode createDiscoveryNode(int port, VirtualMachineScaleSetVM vm,
```
Comment (juan-palacios) [inline src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: This code already has `buildDiscoveredNode`, right? Perhaps we should be consistent in the naming?

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java (hunk 1 above)
Comment (juan-palacios) [reply src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: +1 this is also in a loop so it's getting invoked twice for every element in the loop

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java (hunk 2 above)
Comment (justinskariah) [reply src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: Good point, I will keep it.

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java (hunk 4 above)
Comment (justinskariah) [reply src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: Good point. I will rename it

File: src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java (hunk 3 above)
Comment (justinskariah) [reply src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java]: It's a bit tricky if a scale set has multiple network interface then one of them will be primary and then we can retrieve the primary interface.  We do that in the `if` block. If a scale set has only one network interface then `VirtualMachineScaleSetVM.primaryNetworkInterfaceId()` returns `null` and then we have to iterate through the network interfaces and pick the first one. Hope this makes sense.

**Review by tnymlr — APPROVED** [review]

**Pull request merged**